gradient/src/main.rs

```diff
@@ -0,0 +1,164 @@
+use std::cell::RefCell;
+use std::collections::HashMap;
+use std::io::{BufReader, BufWriter};
+
+use anyhow::Result;
+use elevation::GeoTiffElevation;
+use fs_err::File;
+use geo::{Coord, HaversineLength, LineString};
+use geojson::{Feature, FeatureWriter, Geometry};
+use indicatif::{ParallelProgressIterator, ProgressBar, ProgressStyle};
+use osm_reader::{Element, NodeID, WayID};
+use rayon::prelude::*;
+
+/// This takes an osm.pbf file and a GeoTIFF file covering the area. The GeoTIFF must be a digital
+/// elevation model in WGS84 and units of meters. The output is a GeoJSON file with LineStrings for
+/// every way, with a `gradient` property.
+///
+/// Gradient is:
+///
+/// - positive when the edge is uphill from start to end based on the LineString order, and
+///   negative when downhill
+/// - expressed as an integer percent multiplied by 100 -- a value of `305` means a `3.05%` grade.
+fn main() {
+    let args: Vec<String> = std::env::args().collect();
+    if args.len() != 3 {
+        panic!("Call with the input path to an osm.pbf and a GeoTIFF file");
+    }
+
+    run(&args[1], &args[2], "tmp/gradient.geojson").unwrap();
+}
+
+fn run(osm_path: &str, tif_path: &str, output_path: &str) -> Result<()> {
+    let edges = scrape_osm(osm_path)?;
```

Comment:
Reading OSM takes about 3 mins total for me, and gradient calculations around 5 mins, plus about 30s for writing the final output